src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/DoNotIgnoreMethodResults.cs

```diff
@@ -110,68 +110,193 @@ public override void Initialize(AnalysisContext analysisContext)
 
-                osContext.RegisterOperationAction(opContext =>
+                compilationContext.RegisterOperationBlockStartAction(osContext =>
                 {
-                    IOperation expression = ((IExpressionStatementOperation)opContext.Operation).Operation;
-                    DiagnosticDescriptor rule = null;
-                    string targetMethodName = null;
-                    switch (expression.Kind)
+                    var method = osContext.OwningSymbol as IMethodSymbol;
+                    if (method == null)
                     {
-                        case OperationKind.ObjectCreation:
-                            IMethodSymbol ctor = ((IObjectCreationOperation)expression).Constructor;
-                            if (ctor != null)
-                            {
-                                rule = ObjectCreationRule;
-                                targetMethodName = ctor.ContainingType.Name;
-                            }
-                            break;
-
-                        case OperationKind.Invocation:
-                            IInvocationOperation invocationExpression = ((IInvocationOperation)expression);
-                            IMethodSymbol targetMethod = invocationExpression.TargetMethod;
-                            if (targetMethod == null)
-                            {
-                                break;
-                            }
-
-                            if (IsStringCreatingMethod(targetMethod))
-                            {
-                                rule = StringCreationRule;
-                            }
-                            else if (IsTryParseMethod(targetMethod))
-                            {
-                                rule = TryParseRule;
-                            }
-                            else if (IsHResultOrErrorCodeReturningMethod(targetMethod))
-                            {
-                                rule = HResultOrErrorCodeRule;
-                            }
-                            else if (IsPureMethod(targetMethod, opContext.Compilation))
-                            {
-                                rule = PureMethodRule;
-                            }
-
-                            targetMethodName = targetMethod.Name;
-                            break;
+                        return;
                     }
 
-                    if (rule != null)
+                    osContext.RegisterOperationAction(opContext =>
                     {
-                        Diagnostic diagnostic = Diagnostic.Create(rule, expression.Syntax.GetLocation(), method.Name, targetMethodName);
-                        opContext.ReportDiagnostic(diagnostic);
-                    }
-                }, OperationKind.ExpressionStatement);
+                        if (ShouldSkipAnalyzing(opContext, expectedExceptionType))
```

Comment:
>ShouldSkipAnalyzing [](start = 28, length = 19)

Should we invoke this method after we have determined the method result is unused (just before reporting the diagnostic)? That might be more efficient..